src/main/java/gregtech/api/capability/IFilter.java

```diff
@@ -0,0 +1,144 @@
+package gregtech.api.capability;
+
+import javax.annotation.Nonnull;
+import java.util.Comparator;
+import java.util.function.Predicate;
+
+/**
+ * Base type for generic filters. In addition to the predicate method, this interface provides priority primarily used
+ * in insertion logic.
+ * <p>
+ * Although this class can be used as lambda interface, it is strongly encouraged to define priority according to
+ * criteria specified in javadocs on {@link #getPriority()}.
+ *
+ * @param <T> type of the instance for filtering
+ */
+@FunctionalInterface
+public interface IFilter<T> extends Predicate<T> {
+
+    /**
+     * Compare logic for filter instances.
+     */
+    Comparator<IFilter<?>> FILTER_COMPARATOR = Comparator
+            .<IFilter<?>>comparingInt(IFilter::getPriority)
+            .reversed();
+
+    /**
+     * Test if this filter accepts the instance. If the parameter is {@code null}, then the behavior of this method is
+     * undefined.
+     *
+     * @param t instance to be tested
+     * @return if this filter accepts the instance
+     */
+    @Override
+    boolean test(@Nonnull T t);
+
+    /**
+     * Return insertion priority for this filter. The priority is applied on some insertion logics, to prioritize
+     * certain filters from others. The priority system uses <i>reverse ordering</i>; higher priority values have
+     * precedence over lower ones.
+     * <ul>
+     *     <li>Whitelist filters have {@code Integer.MAX_VALUE - whitelistSize} as their default priority. The lowest
```

Comment:
Do you mean `highest` here? The wording is difficult to determine. Are you referring to the lowest integer value, or the lowest position in the priority?